### PR TITLE
Event enum

### PR DESCRIPTION
**Pull request by SchrodingersGat**
Enumerate available "events" as Enum values, rather than raw strings:

- Ensure that the values can be accurately registered 
- Ability to introspect on available event names
- Automatically document available events
- Ensure plugins keep up to date with event names

### TODO

- [x] Update documentation
- [x] Add unit testing (ensure correct events are triggered)
- [x] Work with python 3.9

### CODE REVIEW

File: src/backend/InvenTree/build/test_build.py [hunk 1]
```diff
@@ -722,6 +724,57 @@ def test_new_build_notification(self):
...
 
+    def test_events(self):
+        """Test that build events are triggered correctly."""
+
+        from django_q.models import OrmQ
+        from build.events import BuildEvents
+
+        settings.TESTING_TABLE_EVENTS = True
+        settings.PLUGIN_TESTING_EVENTS = True
+        set_global_setting('ENABLE_PLUGINS_EVENTS', True)
```
Comment (matmair) [inline src/backend/InvenTree/build/test_build.py:735]: Q: Could a context be used to ensure this setup and teardown are always done correctly?

File: src/backend/InvenTree/build/test_build.py [hunk 2]
```diff
@@ -722,6 +724,57 @@ def test_new_build_notification(self):
...
+        task = findOffloadedEvent(
+            BuildEvents.ISSUED,
+            matching_kwargs=['id'],
+            clear_after=True,
+        )
+
+        self.assertIsNotNone(task)
+
+        settings.TESTING_TABLE_EVENTS = False
+        settings.PLUGIN_TESTING_EVENTS = False
```
Comment (matmair) [inline src/backend/InvenTree/build/test_build.py:776]: This teardown could not be run if there are issues with the asserts above and cause contagion with other tests.

File: src/backend/InvenTree/part/events.py [hunk 3]
```diff
@@ -0,0 +1,7 @@
+"""Event definitions and triggers for the part app."""
+
+from generic.events import BaseEventEnum
+
+
+class PartEvents(BaseEventEnum):
+    """Event enumeration for the Part models."""
```
Comment (matmair) [inline src/backend/InvenTree/part/events.py:7]: Q (blocking): Is this left empty by choice?

File: src/backend/InvenTree/stock/models.py [hunk 4]
```diff
@@ -1255,7 +1256,14 @@ def return_from_customer(self, location, user=None, **kwargs):
             location=location,
         )
 
-        trigger_event('stockitem.returnedfromcustomer', id=self.id)
+        # Clear out allocation information for the stock item
+        self.customer = None
+        self.belongs_to = None
+        self.sales_order = None
+        self.location = location
+        self.clearAllocations()
```
Comment (matmair) [inline src/backend/InvenTree/stock/models.py:1264]: Q (non-blocking): is this intended to be in this diff?

File: src/backend/InvenTree/users/authentication.py
Comment (matmair) [inline src/backend/InvenTree/users/authentication.py:1]: Q (non-blocking): Is this intended to be in this diff?

**Review by matmair — APPROVED** [review]

File: src/backend/InvenTree/part/events.py (hunk 3 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/part/events.py:7]: There are no "part" events yet

File: src/backend/InvenTree/users/authentication.py
Comment (SchrodingersGat) [reply src/backend/InvenTree/users/authentication.py:1]: I was having some issues with circular imports which this improved

**Review by matmair — APPROVED** [review]

File: src/backend/InvenTree/build/test_build.py (hunk 1 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/build/test_build.py:735]: Will do

File: src/backend/InvenTree/build/test_build.py (hunk 2 above)
Comment (SchrodingersGat) [reply src/backend/InvenTree/build/test_build.py:776]: Thanks, I'll fix this

**Pull request merged**